RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc

```diff
@@ -0,0 +1,484 @@
+      for( auto& ahit : thits ) mtdthits.push_back(ahit);
+      thits.swap(mtdthits);
+    }
+    auto trajwithmtd = theTransformer->transform(ttrack,thits);
+    for( const auto& trj : trajwithmtd ) {
+      const auto& thetrj = (updateTraj_ ? trj : trajs.front());
+      reco::Track result = buildTrack(track, thetrj, bs, magfield.product(), !mtdthits.empty());
+      if( result.ndof() > 0 ) {
+	/// setup the track extras
+	reco::TrackExtra::TrajParams trajParams;
+	reco::TrackExtra::Chi2sFive chi2s; 
+	size_t hitsstart = outhits->size();
+	if( updatePattern_ ) {
+	  t2t(trj,*outhits,trajParams,chi2s); // this fills the output hit collection
+	} else {
+	  t2t(thetrj,*outhits,trajParams,chi2s);
+	}
+	size_t hitsend = outhits->size();
+	extras->push_back(buildTrackExtra(trj)); // always push back the fully built extra, update by setting in track
+	extras->back().setHits(hitsRefProd,hitsstart,hitsend);	
+	extras->back().setTrajParams(trajParams,chi2s);
+	//create the track
+	output->push_back(result);
+	auto& backtrack = output->back();
+	reco::TrackExtraRef extraRef(extrasRefProd,extras->size()-1);
+	backtrack.setExtra( (updateExtra_ ? extraRef : track.extra()) );
+	for(unsigned ihit = hitsstart; ihit < hitsend; ++ihit) {
+	  backtrack.appendHitPattern((*outhits)[ihit],ttopo);
+	}
+      }
+    }
+  }
+
+  ev.put(std::move(output));
+  ev.put(std::move(extras));
+  ev.put(std::move(outhits));
+}
+
+template<class TrackCollection>
+TransientTrackingRecHit::ConstRecHitContainer
+TrackExtenderWithMTDT<TrackCollection>::tryBTLLayers(const TrackType& track,
+						     const MTDTrackingDetSetVector& hits,
+						     const MTDDetLayerGeometry* geo,
+						     const MagneticField* field) const {
+  TransientTrackingRecHit::ConstRecHitContainer output;
+  const vector<const DetLayer*>& layers = geo->allBTLLayers();
+  
+  auto cmp = [](const unsigned one, const unsigned two) -> bool { return one < two; };
+
+  auto tTrack = builder->build(track);
+
+  for (auto ilay = layers.begin(); ilay!=layers.end(); ++ilay) {
+    const MTDTrayBarrelLayer* layer = (const MTDTrayBarrelLayer*) (*ilay);
+    
+    // get the outermost trajectory point on the track    
+    TrajectoryStateOnSurface tsos = tTrack.outermostMeasurementState();
+    PropagatorWithMaterial prop(anyDirection,0.13957018,field,1.6,false,0.1,true);
+
+    pair<bool, TrajectoryStateOnSurface> comp = layer->compatible(tsos,prop,*theEstimator);
+    if( comp.first ) {
+      
+      vector<DetLayer::DetWithState> compDets = layer->compatibleDets(tsos,prop,*theEstimator);
+      if (!compDets.empty()) {
+	for( const auto& detWithState : compDets ) {
+	  
+	  auto range = hits.equal_range(detWithState.first->geographicalId(),cmp);	  
+	  for( auto detitr = range.first; detitr != range.second; ++detitr ) {
+	    auto best = detitr->end();
+	    double best_chi2 = std::numeric_limits<double>::max();
+	    for( auto itr = detitr->begin(); itr != detitr->end(); ++itr ) {
```

Comment:
range-based loop (check for `best_chi2 != numeric_limits::max` rather than `best != detitr->end()`)